test/multiple-user-principal-types.test.js

```diff
@@ -0,0 +1,978 @@
+    // override the global app object provided by test/support.js
+    // and create a local one that does not share state with other tests
+    app = loopback({localRegistry: true, loadBuiltinModels: true});
+    app.dataSource('db', {connector: 'memory'});
+
+    // attach User and related models
+    User = app.registry.createModel({
+      name: 'TestUser',
+      base: 'User',
+      // forceId is set to false for the purpose of updating the same affected user within the
+      // `Email Update` test cases.
+      forceId: false,
+      // Speed up the password hashing algorithm for tests
+      saltWorkFactor: 4,
+    });
+    app.model(User, {dataSource: 'db'});
+
+    // create and attach a second User-based model
+    User2 = app.registry.createModel({
+      name: 'TestUser2',
+      base: 'User',
+      // forceId is set to false for the purpose of updating the same affected user within the
+      // `Email Update` test cases.
+      forceId: false,
+      // Speed up the password hashing algorithm for tests
+      saltWorkFactor: 4,
+    });
+    app.model(User2, {dataSource: 'db'});
+
+    AccessToken = app.registry.getModel('AccessToken');
+    app.model(AccessToken, {dataSource: 'db'});
+
+    // Update AccessToken and Users relations to use polymorphism
+    AccessToken.belongsTo('user', {idName: 'id', polymorphic: {idType: 'string',
+      foreignKey: 'userId', discriminator: 'principalType'}});
+    User.hasMany('accessTokens', {polymorphic: {foreignKey: 'userId',
+      discriminator: 'principalType'}});
+    User2.hasMany('accessTokens', {polymorphic: {foreignKey: 'userId',
+      discriminator: 'principalType'}});
+
+    // Speed up the password hashing algorithm
+    // for tests using the built-in User model
+    User.settings.saltWorkFactor = 4;
+
+    // allow many User.afterRemote's to be called
+    User.setMaxListeners(0);
+
+    app.enableAuth({dataSource: 'db'});
+    app.use(loopback.token({model: AccessToken}));
+    app.use(loopback.rest());
+
+    User.create(validCredentials, function(err, user) {
+      if (err) return done(err);
+      validCredentialsUser = user;
+      User2.create(validCredentials, function(err, user2) {
+        if (err) return done(err);
+        validCredentialsUser2 = user2;
+        done(null, user, user2);
+      });
+    });
+  });
+
+  describe('User.login', function() {
+    it('Login a user from 1st user model by providing credentials', function(done) {
+      User.login(validCredentials, function(err, accessToken) {
+        assertGoodToken(accessToken, validCredentialsUser);
+
+        done();
+      });
+    });
+
+    it('Login a user from 2nd user model by providing credentials', function(done) {
+      User2.login(validCredentials, function(err, accessToken) {
+        assertGoodToken(accessToken, validCredentialsUser2);
+
+        done();
+      });
+    });
+
+    it('Login a user by providing email credentials (email case-sensitivity off)', function(done) {
```

Comment:
I think the email case-sensitivity is not related to polymorphic user relation and thus these tests can be removed.